In [1]:
import numpy as np
import pandas as pd

In [5]:
csv_df = pd.read_csv(r"hw_work.csv")
print(csv_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3323 non-null   float64
 7   Total day calls         3323 non-null   float64
 8   Total day charge        3315 non-null   float64
 9   Total eve minutes       3324 non-null   float64
 10  Total eve calls         3325 non-null   float64
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3332 non-null   float64
 14  Total night charge      3333 non-null   

#### DataFrame data
Number of columns is 20

Number of rows is 3333

Total day charge has missing values

The memory usage of the DataFrame is a wee bit over 520.9 KB

In [6]:
print(csv_df['Account length'].describe())


count    3333.000000
mean      101.064806
std        39.822106
min         1.000000
25%        74.000000
50%       101.000000
75%       127.000000
max       243.000000
Name: Account length, dtype: float64


#### Account length data
Min = 1

Max = 243

Mean = 101.064806

Number of rows 3333


In [7]:
print(csv_df.describe().transpose())

                         count        mean        std     min     25%     50%  \
Account length          3333.0  101.064806  39.822106    1.00   74.00  101.00   
Area code               3333.0  437.182418  42.371290  408.00  408.00  415.00   
Number vmail messages   3333.0    8.099010  13.688365    0.00    0.00    0.00   
Total day minutes       3323.0  179.787150  54.419625    0.00  143.70  179.40   
Total day calls         3323.0  100.456214  20.057356    0.00   87.00  101.00   
Total day charge        3315.0   30.557831   9.255987    0.00   24.42   30.50   
Total eve minutes       3324.0  201.033935  50.676652    0.00  166.60  201.40   
Total eve calls         3325.0  100.110677  19.932115    0.00   87.00  100.00   
Total eve charge        3333.0   17.083540   4.310668    0.00   14.16   17.12   
Total night minutes     3333.0  200.872037  50.573847   23.20  167.00  201.20   
Total night calls       3332.0  100.115246  19.566710   33.00   87.00  100.00   
Total night charge      3333

In [8]:
csv_df.tail()

State  Account length  Area code International plan Voice mail plan  \
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3328                     2.0  False  
3329                     3.0  False  
3330                     2.0  False  
3331                     2.0  False  
3332                     0.0  False

In [9]:
csv_df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     26              161.6            123.0   
2                      0              243.4            114.0   
3                      0              299.4             71.0   
4                      0              166.7            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             27.47              195.5            103.0             16.62   
2             41.38              121.2            110.0             10.30   
3             50.90               61.9             88.0              5.26   
4             28.34              148.3            122.0             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7               91.0               11.01   
1                254.4              103.0               11.45   
2                162.6              104.0                7.32   
3                196.9               89.0                8.86   
4                186.9              121.0                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                     1.0  False  
1                     1.0  False  
2                     0.0  False  
3                     2.0  False  
4                     3.0  False

In [10]:
csv_df['State']

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

In [12]:
csv_df[['State','International plan']]

State International plan
0       KS                 No
1       OH                 No
2       NJ                 No
3       OH                Yes
4       OK                Yes
...    ...                ...
3328    AZ                 No
3329    WV                 No
3330    RI                 No
3331    CT                Yes
3332    TN                 No

[3333 rows x 2 columns]

In [19]:
csv_df[csv_df['International plan'] == 'Yes']

State  Account length  Area code International plan Voice mail plan  \
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
5       AL             118        510                Yes              No   
7       MO             147        415                Yes              No   
9       WV             141        415                Yes             Yes   
...    ...             ...        ...                ...             ...   
3304    IL              71        510                Yes              No   
3309    VT             100        408                Yes              No   
3317    SD             163        415                Yes              No   
3320    GA             122        510                Yes              No   
3331    CT             184        510                Yes              No   

      Number vmail messages  Total day minutes  Total day calls  \
3                         0              299.4             71.0   
4                         0              166.7            113.0   
5                         0              223.4             98.0   
7                         0              157.0             79.0   
9                        37              258.6             84.0   
...                     ...                ...              ...   
3304                      0              186.1            114.0   
3309                      0              219.4            112.0   
3317                      0              197.2             90.0   
3320                      0              140.0            101.0   
3331                      0              213.8            105.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
5                37.98              220.6            101.0             18.75   
7                26.69              103.1             94.0              8.76   
9                43.96              222.0            111.0             18.87   
...                ...                ...              ...               ...   
3304             31.64              198.6            140.0             16.88   
3309             37.30              225.7            102.0             19.18   
3317             33.52              188.5            113.0             16.02   
3320             23.80              196.4             77.0             16.69   
3331             36.35              159.6             84.0             13.57   

      Total night minutes  Total night calls  Total night charge  \
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
5                   203.9              118.0                9.18   
7                   211.8               96.0                9.53   
9                   326.4               97.0               14.69   
...                   ...                ...                 ...   
3304                206.5               80.0                9.29   
3309                255.3               95.0               11.49   
3317                211.1               94.0                9.50   
3320                120.1              133.0                5.40   
3331                139.2              137.0                6.26   

      Total intl minutes  Total intl calls  Total intl charge  \
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
5                    6.3                 6               1.70   
7                    7.1                 6               1.92   
9                   11.2                 5               3.02   
...                  ...               ...                ...   
3304                13.8                 5            

In [21]:
csv_df[(csv_df['International plan'] == 'Yes') & (csv_df['Churn'] == True)]

State  Account length  Area code International plan Voice mail plan  \
41      MD             135        408                Yes             Yes   
115     ME              36        510                Yes             Yes   
144     VT             117        408                Yes              No   
198     ME             131        510                Yes             Yes   
214     FL              70        510                Yes              No   
...    ...             ...        ...                ...             ...   
3246    NC              77        408                Yes             Yes   
3255    RI             138        510                Yes              No   
3291    MI             119        510                Yes             Yes   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   

      Number vmail messages  Total day minutes  Total day calls  \
41                       41              173.1             85.0   
115                      42              196.8             89.0   
144                       0              167.1             86.0   
198                      26              292.9            101.0   
214                       0              226.7             98.0   
...                     ...                ...              ...   
3246                     44              103.2            117.0   
3255                      0              286.2             61.0   
3291                     22              172.1            119.0   
3304                      0              186.1            114.0   
3320                      0              140.0            101.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
41               29.43              203.9            107.0             17.33   
115              33.46              254.9            122.0             21.67   
144              28.41              177.5             87.0             15.09   
198              49.79              199.7             97.0             16.97   
214              38.54              228.1            115.0             19.39   
...                ...                ...              ...               ...   
3246             17.54              236.3             86.0             20.09   
3255             48.65              187.2             60.0             15.91   
3291             29.26              223.6            133.0             19.01   
3304             31.64              198.6            140.0             16.88   
3320             23.80              196.4             77.0             16.69   

      Total night minutes  Total night calls  Total night charge  \
41                  122.2               78.0                5.50   
115                 138.3              126.0                6.22   
144                 249.4              132.0               11.22   
198                 255.3              127.0               11.49   
214                  73.2               93.0                3.29   
...                   ...                ...                 ...   
3246                203.5              101.0                9.16   
3255                146.2              114.0                6.58   
3291                150.0               94.0                6.75   
3304                206.5               80.0                9.29   
3320                120.1              133.0                5.40   

      Total intl minutes  Total intl calls  Total intl charge  \
41                  14.6                15               3.94   
115                 20.0                 6               5.40   
144                 14.1                 7               3.81   
198                 13.8                 7               3.73   
214                 17.6                 4               4.75   
...                  ...               ...                ...   
3246                11.9                 2            

In [22]:
csv_df.iloc[14]

State                        IA
Account length               62
Area code                   415
International plan           No
Voice mail plan              No
Number vmail messages         0
Total day minutes           NaN
Total day calls             NaN
Total day charge            NaN
Total eve minutes         307.2
Total eve calls            76.0
Total eve charge          26.11
Total night minutes       203.0
Total night calls          99.0
Total night charge         9.14
Total intl minutes         13.1
Total intl calls              6
Total intl charge          3.54
Customer service calls      4.0
Churn                     False
Name: 14, dtype: object

In [23]:
csv_df.iloc[14:]

State  Account length  Area code International plan Voice mail plan  \
14      IA              62        415                 No              No   
15      NY             161        415                 No              No   
16      ID              85        408                 No             Yes   
17      VT              93        510                 No              No   
18      VA              76        510                 No             Yes   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
14                        0                NaN              NaN   
15                        0                NaN              NaN   
16                       27                NaN              NaN   
17                        0                NaN              NaN   
18                       33                NaN              NaN   
...                     ...                ...              ...   
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
14                 NaN              307.2             76.0             26.11   
15                 NaN              317.8             97.0             27.01   
16                 NaN              280.9             90.0             23.88   
17                 NaN              218.2            111.0             18.55   
18                 NaN              212.8             65.0             18.09   
...                ...                ...              ...               ...   
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
14                  203.0               99.0                9.14   
15                  160.6              128.0                7.23   
16                   89.3                NaN                4.02   
17                  129.6              121.0                5.83   
18                  165.7              108.0                7.46   
...                   ...                ...                 ...   
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
14                  13.1                 6               3.54   
15                   5.4                 9               1.46   
16                  13.8                 4               3.73   
17                   8.1                 3               2.19   
18                  10.0                 5               2.70   
...                  ...               ...                ...   
3328                 9.9                 6            

In [28]:
csv_df.set_index('State')

Account length  Area code International plan Voice mail plan  \
State                                                                 
KS                128        415                 No             Yes   
OH                107        415                 No             Yes   
NJ                137        415                 No              No   
OH                 84        408                Yes              No   
OK                 75        415                Yes              No   
...               ...        ...                ...             ...   
AZ                192        415                 No             Yes   
WV                 68        415                 No              No   
RI                 28        510                 No              No   
CT                184        510                Yes              No   
TN                 74        415                 No             Yes   

       Number vmail messages  Total day minutes  Total day calls  \
State                                                              
KS                        25              265.1            110.0   
OH                        26              161.6            123.0   
NJ                         0              243.4            114.0   
OH                         0              299.4             71.0   
OK                         0              166.7            113.0   
...                      ...                ...              ...   
AZ                        36              156.2             77.0   
WV                         0              231.1             57.0   
RI                         0              180.8            109.0   
CT                         0              213.8            105.0   
TN                        25              234.4            113.0   

       Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
State                                                                           
KS                45.07              197.4             99.0             16.78   
OH                27.47              195.5            103.0             16.62   
NJ                41.38              121.2            110.0             10.30   
OH                50.90               61.9             88.0              5.26   
OK                28.34              148.3            122.0             12.61   
...                 ...                ...              ...               ...   
AZ                26.55              215.5            126.0             18.32   
WV                39.29              153.4             55.0             13.04   
RI                30.74              288.8             58.0             24.55   
CT                36.35              159.6             84.0             13.57   
TN                39.85              265.9             82.0             22.60   

       Total night minutes  Total night calls  Total night charge  \
State                                                               
KS                   244.7               91.0               11.01   
OH                   254.4              103.0               11.45   
NJ                   162.6              104.0                7.32   
OH                   196.9               89.0                8.86   
OK                   186.9              121.0                8.41   
...                    ...                ...                 ...   
AZ                   279.1               83.0               12.56   
WV                   191.3              123.0                8.61   
RI                   191.9               91.0                8.64   
CT                   139.2              137.0                6.26   
TN                   241.4               77.0               10.86   

       Total intl minutes  Total intl calls  Total intl charge  \
State                                                            
KS                   10.0                 3               2.70   
OH                   13.7                 3     

In [34]:
try:
    row = csv_df.loc['LA']
    print(row)
except KeyError:
    print("Index 'LA' not found.")

Index 'LA' not found.


In [37]:
csv_df.drop(['Area code'], axis = 1)

State  Account length International plan Voice mail plan  \
0       KS             128                 No             Yes   
1       OH             107                 No             Yes   
2       NJ             137                 No              No   
3       OH              84                Yes              No   
4       OK              75                Yes              No   
...    ...             ...                ...             ...   
3328    AZ             192                 No             Yes   
3329    WV              68                 No              No   
3330    RI              28                 No              No   
3331    CT             184                Yes              No   
3332    TN              74                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6              

In [39]:
csv_df.drop([3328])

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3327    SC              79        415                 No              No   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3327                      0              134.7             98.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3327             22.90              189.7             68.0             16.12   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3327                221.4              128.0                9.96   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3327                11.8                 5            

In [41]:
csv_df['total_min'] = csv_df['Total night minutes'] + csv_df['Total intl minutes']
csv_df['mean_min'] = (csv_df['Total night minutes'] + csv_df['Total intl minutes'])/2

csv_df

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  ...  Total night minutes  \
0                45.07              197.4  ...                244.7   
1                27.47              195.5  ...                254.4   
2                41.38              121.2  ...                162.6   
3                50.90               61.9  ...                196.9   
4                28.34              148.3  ...                186.9   
...                ...                ...  ...                  ...   
3328             26.55              215.5  ...                279.1   
3329             39.29              153.4  ...                191.3   
3330             30.74              288.8  ...                191.9   
3331             36.35              159.6  ...                139.2   
3332             39.85              265.9  ...                241.4   

      Total night calls  Total night charge  Total intl minutes  \
0                  91.0               11.01                10.0   
1                 103.0               11.45                13.7   
2                 104.0                7.32                12.2   
3                  89.0                8.86                 6.6   
4                 121.0                8.41                10.1   
...                 ...                 ...                 ...   
3328               83.0               12.56                 9.9   
3329              123.0                8.61                 9.6   
3330               91.0                8.64                14.1   
3331              137.0                6.26                 5.0   
3332               77.0               10.86                13.7   

      Total intl calls  Total intl charge  Customer service calls  Churn  \
0                    3               2.70                     1.0  False   
1                    3               3.70                     1.0  False   
2                    5               3.29                     0.0  False   
3                    7               1.78                     2.0  False   
4                    3               2.73                     3.0  False   
...                ...                ...                     ...    ...   
3328                 6               2.67                     2.0  False   
3329                 

In [47]:
new_df = pd.concat([csv_df.head(1), csv_df.tail(1)], ignore_index = True)
# new_df.reset_index(drop=True, inplace=True)
print(f'new_df:\n{new_df}')

new_df:
  State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    TN              74        415                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     25              234.4            113.0   

   Total day charge  Total eve minutes  ...  Total night minutes  \
0             45.07              197.4  ...                244.7   
1             39.85              265.9  ...                241.4   

   Total night calls  Total night charge  Total intl minutes  \
0               91.0               11.01                10.0   
1               77.0               10.86                13.7   

   Total intl calls  Total intl charge  Customer service calls  Churn  \
0                 3                2.7                     1.0  False   
1                 4                

In [48]:
row = {'State': 'LA', 'Account length': 124, 'Area code': 64, 'International plan': 'Yes', 'Voice mail plan':'Yes', 'Number vmail messages':125, 'Total day minutes':300, 'Total day calls':150, 'Total day charge':45, 'Total eve minutes':17.2, 'Total eve calls':76.0, 'Total eve charge':26.11, 'Total night minutes':203.0, 'Total night calls':99.0, 'Total night charge':9.14, 'Total intl minutes':13.1, 'Total intl calls':6, 'Total intl charge':3.54, 'Customer service calls':4.0, 'Churn':True}
new_row = pd.DataFrame([row])
pd.concat([new_df, pd.DataFrame([row])], ignore_index=True).tail()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    TN              74        415                 No             Yes   
2    LA             124         64                Yes             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     25              234.4            113.0   
2                    125              300.0            150.0   

   Total day charge  Total eve minutes  ...  Total night minutes  \
0             45.07              197.4  ...                244.7   
1             39.85              265.9  ...                241.4   
2             45.00               17.2  ...                203.0   

   Total night calls  Total night charge  Total intl minutes  \
0               91.0               11.01                10.0   
1               77.0               10.86                13.7   
2               99.0                9.14                13.1   

   Total intl calls  Total intl charge  Customer service calls  Churn  \
0                 3               2.70                     1.0  False   
1                 4               3.70                     0.0  False   
2                 6               3.54                     4.0   True   

   total_min mean_min  
0      254.7   127.35  
1      255.1   127.55  
2        NaN      NaN  

[3 rows x 22 columns]

#### 17
The inplace parameter inpandas methods determines whether the operation modifies the original object directly or creates and returns a new modified object.

 - Use inplace=False for better code readability and to avoid accidental changes to the original object.
 - Use inplace=True when you need to save memory or explicitly want to modify the object.

#### 18
When you set a column as the index using set_index(), it becomes the index and is removed from the main DataFrame columns.

based on: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html

#### 19
axis = 0 means columns

axis = 1 means rows


#### 20
- & (AND operator): Used to combine multiple conditions where both must be true.
- | (OR operator): Used to combine conditions where at least one must be true.
- ~ (NOT operator): Negates a condition, flipping True to False and vice versa.
- isin (Check if value exists in a list or set): Tests if values in a column exist in a specified list or set.
- between (Check if values fall within a range): Tests if values in a column are within a range (inclusive by default).